js/render.js

```diff
@@ -11,26 +11,172 @@ function drawBlock( x, y ) {
+	        for ( var x = 0; x < 4; ++x ) {
+	            if ( current[ y ][ x ] ) {
+	                if( !checkLast() ) {
+	                    ctx.fillStyle = colors[ current[ y ][ x ] - 1 ];
+	                    drawBlock( currentX + x, currentY + y );
+	                    ctx.globalAlpha = 0.4;
+	                    drawBlock( ghostCurrentX + x, ghostCurrentY + y );
+	                    ctx.globalAlpha = 1.0;
+	                }
+	                else {
+	                    if( !lose ) {
+	                        drawLast();
+	                    }
+	                    return;
+	                }
+	            }
+	        }
+	    }
+	}
+}
+
+function drawLast() {
+    ctx.strokeStyle = 'black';
+    var row = 10;
+    var cnt = 0;
+    for ( var y = 3; y >= 0 ; --y ) {
+        for ( var x = 0; x < 4; ++x ) {
+            if ( current[ y ][ x ] ) {
+                for (var r = 2; r>=0; --r ) {
+                    if( !board[ r ][ x + currentX ] ){
+                        if ( row > r + cnt ) {
+                            row = r + cnt;
+                        }
+                    }
+                }
+            }
+        }
+        if( row != 10 ){
+            cnt++;
+        }
+    }
+    if( row == 10 ){
+        return;
+    }
+
+    var cnt = 0;
+    var empty = 1;
+    for ( var y = 3; y >= 0; y-- ) {
+        if ( row - cnt < 0 ) {
+            break;
+        }
+        for ( var x = 3; x >= 0; x-- ) {
+            if ( current[ y ][ x ] ) {
+                if ( typeof board[ row - cnt ] == 'undefined'
+                    || typeof board[ row - cnt ][ x + currentX ] == 'undefined'
+                    || !board[ row - cnt][ currentX + x ] ) {
+                        ctx.fillStyle = colors[ current[ y ][ x ] - 1 ];
+                        board[ row - cnt ][ currentX + x ] = current[ y ][ x ];
+                        drawBlock( currentX + x, row - cnt );
+                        empty = 0;
+                }
+            }
+        }
+        if(!empty){
+            cnt++;
+        }
+    }
+}
+
+function drawLine() {
+    ctx.clearRect( 300, 0, 140, 600 );
     ctx.strokeStyle = 'black';
+    ctx.beginPath();
+    ctx.moveTo( 300, 0 );
+    ctx.lineTo( 300, 600 );
+    ctx.stroke();
+    ctx.font = '18.5pt Arial';
+    ctx.fillStyle = 'black';
+    ctx.fillText( 'next block', 315, 40 );
+}
+
+function drawNext() {
+    ctx.clearRect( 301, 50, 139, 350 );
+    for ( var y = 0; y < 4; ++y ) {
+        for ( var x = 0; x < 4; ++x ) {
+            if ( next1[ y ][ x ] ) {
+                ctx.fillStyle = colors[ next1[ y ][ x ] - 1 ];
+                drawBlock( 10.5 + x, 2.5 + y );
+            }
+            if ( next2[ y ][ x ] ) {
+                ctx.fillStyle = colors[ next2[ y ][ x ] - 1 ];
+                drawBlock( 10.5 + x, 6 + y );
+            }
+            if ( next3[ y ][ x ] ) {
+                ctx.fillStyle = colors[ next3[ y ][ x ] - 1 ];
+                drawBlock( 10.5 + x, 9.5 + y );
```

Comment:
Separate constants into beginning of file, make them semantic

Reply:
Please check [our final resort](https://18-1-skku-oss.github.io/canvas-tetris/) to see what's going on! We are planning to clean up, simplify, and also optimize much more for better code.